File: myrtle/befor0225/palindrome_linked_list.py
# https://leetcode.com/problems/palindrome-linked-list/

# 最简单的方法当然是把这条链倒转，然后比较新链跟旧链

# 但是慢，用 stack 的话，最大的问题是对称性的问题，局部对称性推不出整体对称性。

# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def isPalindrome(self, head: ListNode) -> bool:
        return None

# 看了答案，结果还是要 reverse 啊

File: myrtle/goal/src/goal/palindrome_linked_list.go
package goal

func isPalindrome(head *ListNode) bool {
	data := []int{}
	for {
		if head == nil {
			break
		}
		data = append(data, head.Val)
		if head.Next == nil {
			break
		}
		head = head.Next
	}
	if len(data) <= 1 {
		return true
	}
	h, t := 0, len(data)-1
	for {
		if h >= t {
			return true
		}
		if data[h] != data[t] {
			return false
		}
		h += 1
		t -= 1
	}
}
